### PR TITLE
Context menus migration guide

### PR DESCRIPTION
**Pull request by justinmc**
A migration guide for the Customizable Context Menus feature (https://github.com/flutter/flutter/pull/107193).

### CODE REVIEW

File: src/release/breaking-changes/context-menus.md [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+added to all text-editing and text-selection widgets. If one is not provided,
+then Flutter just sets it to a default that builds the correct context menu for
+the given platform. All of these default widgets are exposed to users for
+re-use. Customizing context menus now consists of using contextMenuBuilder to
+return whatever widget you want, possibly including reusing the built-in context
+menu widgets.
+
+Here's an example that shows how to add a "Send email" button to the default
+context menus whenever an email address is selected. The full code can be found
+[here](https://github.com/flutter/samples/blob/main/experimental/context_menus/lib/email_button_page.dart).
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:32]:
```suggestion
in [email_button_page.dart]({{site.github}}/flutter/samples/blob/main/experimental/context_menus/lib/email_button_page.dart)
on GitHub.
```

Never never never link on "here".  :D

File: src/release/breaking-changes/context-menus.md [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    return AdaptiveTextSelectionToolbar.buttonItems(
+      anchors: editableTextState.contextMenuAnchors,
+      buttonItems: buttonItems,
+    );
+  },
+)
+```
+
+A large number of examples of different custom context menus are available [in
+the samples repo](https://github.com/flutter/samples/tree/main/experimental/context_menus).
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:60]:
```suggestion
the samples repo]({{site.github}}/flutter/samples/tree/main/experimental/context_menus).
```

File: src/release/breaking-changes/context-menus.md [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+All related deprecated features were flagged with the deprecation warning "Use
+`contextMenuBuilder` instead."
+
+## Migration guide
+In general, any previous changes to context menus that have been deprecated now
+require the use of the contextMenuBuilder parameter on the relevant text-editing
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:67]:
```suggestion
require the use of the `contextMenuBuilder` parameter on the relevant text-editing
```

File: src/release/breaking-changes/context-menus.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+or text-selection widget (
+[on TextField](https://api.flutter.dev/flutter/material/TextField/contextMenuBuilder.html),
+for example). Return a built-in context menu widget like
+[AdaptiveTextSelectionToolbar](https://api.flutter.dev/flutter/material/AdaptiveTextSelectionToolbar-class.html)
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:71]:
```suggestion
[on `TextField`]({{site.api}}/flutter/material/TextField/contextMenuBuilder.html),
for example). Return a built-in context menu widget like
[`AdaptiveTextSelectionToolbar`]({{site.api}}/flutter/material/AdaptiveTextSelectionToolbar-class.html)
```

File: src/release/breaking-changes/context-menus.md [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+to use Flutter's built-in context menus, or return your own widget for something
+totally custom.
+
+To transition to contextMenuBuilder, the following parameters and classes have
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:75]:
```suggestion
To transition to `contextMenuBuilder`, the following parameters and classes have
```

File: src/release/breaking-changes/context-menus.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+been deprecated.
+
+### [ToolbarOptions](https://api.flutter.dev/flutter/widgets/ToolbarOptions-class.html)
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:78]:
```suggestion
### [ToolbarOptions]({{site.api}}/flutter/widgets/ToolbarOptions-class.html)
```

File: src/release/breaking-changes/context-menus.md [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+This class was previously used to explicitly enable or disable certain buttons
+in a context menu. Before this change, you might have passed it into TextField
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:80]:
```suggestion
in a context menu. Before this change, you might have passed it into `TextField`
```

File: src/release/breaking-changes/context-menus.md [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+TextField(
+  toolbarOptions: ToolbarOptions(
+    copy: true,
+  ),
+)
+```
+
+Now, you can achieve the same effect by adjusting the buttonItems passed into
+AdaptiveTextSelectionToolbar. For example, you could ensure that the Cut button
+never appears, but the other buttons do appear as usual:
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:94]:
```suggestion
Now, you can achieve the same effect by adjusting the `buttonItems` passed into
`AdaptiveTextSelectionToolbar`. For example, you could ensure that the **Cut** button
never appears, but the other buttons do appear as usual:
```

File: src/release/breaking-changes/context-menus.md [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    });
+    return AdaptiveTextSelectionToolbar.buttonItems(
+      anchors: editableTextState.contextMenuAnchors,
+      buttonItems: buttonItems,
+    );
+  },
+)
+```
+
+Or, you could ensure that the Cut button appears exclusively and always:
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:112]:
```suggestion
Or, you could ensure that the **Cut** button appears exclusively and always:
```

File: src/release/breaking-changes/context-menus.md [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+          },
+          type: ContextMenuButtonType.cut,
+        ),
+      ],
+    );
+  },
+)
+```
+
+### [TextSelectionControls.canCut](https://api.flutter.dev/flutter/widgets/TextSelectionControls/canCut.html) and other button booleans
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:132]:
```suggestion
### [TextSelectionControls.canCut]({{site.api}}/flutter/widgets/TextSelectionControls/canCut.html) and other button booleans
```

File: src/release/breaking-changes/context-menus.md [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+These booleans previously had the same effect of enabling and disabling certain
+buttons as ToolbarOptions.cut, etc. had. Before this change, you might have been
+hiding and showing buttons by overriding TextSelectionControls and setting these
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:135]:
```suggestion
buttons as `ToolbarOptions.cut`, and so on had. Before this change, you might have been
hiding and showing buttons by overriding `TextSelectionControls` and setting these
```

File: src/release/breaking-changes/context-menus.md [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+```dart
+// Deprecated.
+class _MyMaterialTextSelectionControls extends MaterialTextSelectionControls {
+  @override
+  bool canCut() => false,
+}
+```
+
+See the previous section on ToolbarOptions for how to achieve a similar effect
+with contextMenuBuilder.
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:147]:
```suggestion
See the previous section on `ToolbarOptions` for how to achieve a similar effect
with `contextMenuBuilder`.
```

File: src/release/breaking-changes/context-menus.md [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+
+### [TextSelectionControls.handleCut](https://api.flutter.dev/flutter/widgets/TextSelectionControls/handleCut.html) and other button callbacks
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:149]:
```suggestion
### [TextSelectionControls.handleCut]({{site.api}}/flutter/widgets/TextSelectionControls/handleCut.html) and other button callbacks
```

File: src/release/breaking-changes/context-menus.md [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+  @override
+  bool handleCut() {
+    // My custom cut implementation here.
+  },
+}
+```
+
+This is still possible using contextMenuBuilder, including calling
+out to the original buttons' actions in the custom handler, using toolbar
+widgets like AdaptiveTextSelectionToolbar.buttonItems.
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:166]:
```suggestion
This is still possible using `contextMenuBuilder`, including calling
out to the original buttons' actions in the custom handler, using toolbar
widgets like `AdaptiveTextSelectionToolbar.buttonItems`.
```

File: src/release/breaking-changes/context-menus.md [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+
+This example shows modifying the Copy button to show a dialog in addition to
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:168]:
```suggestion
This example shows modifying the **Copy** button to show a dialog in addition to
```

File: src/release/breaking-changes/context-menus.md [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+      anchors: editableTextState.contextMenuAnchors,
+      buttonItems: buttonItems,
+    );
+  },
+)
+```
+
+A full example of modifying a built-in context menu action can be found in the
+samples repository
+[here](https://github.com/flutter/samples/blob/main/experimental/context_menus/lib/modified_action_page.dart).
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:209]:
```suggestion
samples repository, in [email_button_page.dart]
({{site.github}}/flutter/samples/blob/main/experimental/context_menus/lib/modified_action_page.dart).
```

File: src/release/breaking-changes/context-menus.md [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+
+### [buildToolbar](https://api.flutter.dev/flutter/widgets/TextSelectionControls/buildToolbar.html)
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:211]:
```suggestion
### [`buildToolbar`]({{site.api}}/flutter/widgets/TextSelectionControls/buildToolbar.html)
```

File: src/release/breaking-changes/context-menus.md [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+This function generated the context menu widget similarly to contextMenuBuilder,
+but required much more setup to work with. Before this change, you might have
+been overriding buildToolbar as a part of TextSelectionControls, like this:
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:214]:
```suggestion
This function generated the context menu widget similarly to `contextMenuBuilder`,
but required more setup to use. Before this change, you might have
been overriding `buildToolbar` as a part of `TextSelectionControls`, like this:
```

File: src/release/breaking-changes/context-menus.md [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+    Offset lastSecondaryTapDownPosition,
+  ) {
+    return _MyCustomToolbar();
+  },
+}
+```
+
+Now you can simply use contextMenuBuilder directly as a parameter to TextField
+(and others). The information provided in the parameters to buildToolbar can be
+obtained via the EditableTextState that is passed to contextMenuBuilder.
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:237]:
```suggestion
Now you can simply use `contextMenuBuilder` directly as a parameter to `TextField`
(and others). The information provided in the parameters to `buildToolbar` can be
obtained from the `EditableTextState` that is passed to `contextMenuBuilder`.
```

File: src/release/breaking-changes/context-menus.md [hunk 20]
```diff
@@ -0,0 +1,328 @@
...
+        );
+      },
+    );
+  }
+}
+```
+
+A full example of building a custom context menu can be found in the samples
+repository
+[here](https://github.com/flutter/samples/blob/main/experimental/context_menus/lib/custom_menu_page.dart).
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:298]:
```suggestion
repository, [custom_menu_page.dart]
({{site.github}}/flutter/samples/blob/main/experimental/context_menus/lib/custom_menu_page.dart).
```

File: src/release/breaking-changes/context-menus.md [hunk 21]
```diff
@@ -0,0 +1,328 @@
...
+Relevant issues:
+
+* [Simple custom text selection toolbars] {{site.repo.flutter}}/issues/73574
+* [Right click menu outside of text fields] {{site.repo.flutter}}/issues/98272
+* [Text editing for desktop - stable] {{site.repo.flutter}}/issues/90563
+* [Ability to disable context menu on TextFields] {{site.repo.flutter}}/issues/79796
+* [Missing APIs for text selection toolbar styling] {{site.repo.flutter}}/issues/22210
+* [Enable copy toolbar in all widgets] {{site.repo.flutter}}/issues/49996
+* [Disable context menu from broser] {{site.repo.flutter}}/issues/78671
+* [Custom context menus don't show up for Flutter web] {{site.repo.flutter}}/issues/84219
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:321]:
```suggestion
* [Simple custom text selection toolbars]({{site.repo.flutter}}/issues/73574)
* [Right click menu outside of text fields]({{site.repo.flutter}}/issues/98272)
* [Text editing for desktop - stable]({{site.repo.flutter}}/issues/90563)
* [Ability to disable context menu on TextFields]({{site.repo.flutter}}/issues/79796)
* [Missing APIs for text selection toolbar styling]({{site.repo.flutter}}/issues/22210)
* [Enable copy toolbar in all widgets]({{site.repo.flutter}}/issues/49996)
* [Disable context menu from browser]({{site.repo.flutter}}/issues/78671)
* [Custom context menus don't show up for Flutter web]({{site.repo.flutter}}/issues/84219)
```

File: src/release/breaking-changes/context-menus.md [hunk 22]
```diff
@@ -0,0 +1,328 @@
...
+
+Relevant PRs:
+
+* [ContextMenus] {{site.repo.flutter}}/pull/107193
+* [Ability to disable the browser's context menu on web] {{site.repo.flutter}}/pull/118194
+* [Ability to disable the browser's context menu on web (engine)] {{site.repo.engine}}/pull/38682
+* [Custom context menus in SelectableRegion on web] {{site.repo.flutter}}/pull/121653
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:328]:
```suggestion
* [ContextMenus]({{site.repo.flutter}}/pull/107193)
* [Ability to disable the browser's context menu on web]({{site.repo.flutter}}/pull/118194)
* [Ability to disable the browser's context menu on web (engine)]({{site.repo.engine}}/pull/38682)
* [Custom context menus in SelectableRegion on web]({{site.repo.flutter}}/pull/121653)
```

File: src/release/breaking-changes/context-menus.md [hunk 23]
```diff
@@ -0,0 +1,328 @@
+---
+title: A new way to customize context menus
+description: Several hard-coded parameters for customizing context menus have now been replaced by a generic widget builder.
+---
+
+## Summary
+Context menus, or text selection toolbars, are the menus that show up when long
+pressing or right clicking on text in Flutter, and they show options like Cut,
+Copy, Paste, and Select all. Previously, it was only possible to narrowly
+customize them using ToolbarOptions and TextSelectionControls. Now, they have
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:10]:
```suggestion
**Copy**, **Paste**, and **Select all**. Previously, it was only possible to narrowly
customize them using `ToolbarOptions` and `TextSelectionControls`. Now, they have
```

File: src/release/breaking-changes/context-menus.md [hunk 24]
```diff
@@ -0,0 +1,328 @@
...
+been made composable using widgets, just like everything else in Flutter, and
+the specific configuration parameters have been deprecated.
+
+## Context
+Previously, it was possible to disable buttons from the context menus using
+TextSelectionControls, but any customization beyond that required copying and
+editing hundreds of lines of custom classes in the framework. Now, all of this
+has been replaced by a simple builder function, contextMenuBuilder, which allows
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:18]:
```suggestion
`TextSelectionControls`, but any customization beyond that required copying and
editing hundreds of lines of custom classes in the framework. Now, all of this
has been replaced by a simple builder function, `contextMenuBuilder`, which allows
```

File: src/release/breaking-changes/context-menus.md [hunk 25]
```diff
@@ -0,0 +1,328 @@
...
+any Flutter widget to be used as a context menu.
+
+## Description of change
+Context menus are now built via the contextMenuBuilder parameter, which has been
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:22]:
```suggestion
Context menus are now built from the `contextMenuBuilder` parameter, which has been
```

File: src/release/breaking-changes/context-menus.md [hunk 26]
```diff
@@ -0,0 +1,328 @@
...
+added to all text-editing and text-selection widgets. If one is not provided,
+then Flutter just sets it to a default that builds the correct context menu for
+the given platform. All of these default widgets are exposed to users for
+re-use. Customizing context menus now consists of using contextMenuBuilder to
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:26]:
```suggestion
re-use. Customizing context menus now consists of using `contextMenuBuilder` to
```

File: src/release/breaking-changes/context-menus.md [hunk 27]
```diff
@@ -0,0 +1,328 @@
...
+return whatever widget you want, possibly including reusing the built-in context
+menu widgets.
+
+Here's an example that shows how to add a "Send email" button to the default
```
Comment (sfshaza2) [inline src/release/breaking-changes/context-menus.md:30]:
```suggestion
Here's an example that shows how to add a **Send email** button to the default
```

**Review by sfshaza2 — COMMENTED** [review]
Lots of little nits. I was going to ignore most, but then you had to go and link on "here". That's a "never never never" situation.  :D

File: src/release/breaking-changes/context-menus.md (hunk 11 above)
Comment (justinmc) [reply src/release/breaking-changes/context-menus.md]: I was really watching my "i.e."s and "e.g."s, but I forgot about "etc." :)

**Comment by justinmc** [comment]
@sfshaza2 Thank you for the review!  I've addressed all your comments.  And linking on "here" has been erased from my vocabulary 🧠

**Review by sfshaza2 — APPROVED** [review]
lgtm. One final nit. Can you add this to the index page? Then it can land.  :D

**Comment by justinmc** [comment]
@sfshaza2 Added to index.md, but can you double check that I linked it correctly?  It looks like most of the links on that page are broken.  Referring to https://github.com/flutter/website/blob/main/src/release/breaking-changes/index.md.

File: src/release/breaking-changes/index.md [hunk 28]
```diff
@@ -31,6 +31,7 @@ release, and listed in alphabetical order:
 * [Migrate `of` to non-nullable return values, and add `maybeOf`][]
 * [Removed RouteSettings.copyWith][]
 * [ThemeData's toggleableActiveColor property has been deprecated][]
+* [A new way to customize context menus]({{site.github}}/flutter/website/blob/main/src/release/breaking-changes/context-menus.md)
```
Comment (sfshaza2) [inline src/release/breaking-changes/index.md:34]:
```suggestion
* [A new way to customize context menus]({{site.url}}/release/breaking-changes/context-menus.md)
```

**Comment by sfshaza2** [comment]
Hi, @justinmc! No, you can't go by that page as it hasn't gone through all of the jekyll/liquid processing. You need to look at the live page: https://docs.flutter.dev/release/breaking-changes.  I've added the correct link to the PR.

**Comment by justinmc** [comment]
Oh got it, thanks!  I'll merge this when it's green then.

File: src/release/breaking-changes/index.md [hunk 29]
```diff
@@ -31,6 +31,7 @@ release, and listed in alphabetical order:
 * [Migrate `of` to non-nullable return values, and add `maybeOf`][]
 * [Removed RouteSettings.copyWith][]
 * [ThemeData's toggleableActiveColor property has been deprecated][]
+* [A new way to customize context menus]({{site.url}}/release/breaking-changes/context-menus.md)
```
Comment (parlough) [inline src/release/breaking-changes/index.md:34]: We want to link to the generated HTML not Markdown:

```suggestion
* [A new way to customize context menus]({{site.url}}/release/breaking-changes/context-menus)
```

**Review by Piinks — APPROVED** [review]
Thank you @justinmc! LGTM!

**Pull request merged**